extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -77,40 +85,129 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         deviceClient.setEndpoint(endPoint);
-        updateStatus(ThingStatus.UNKNOWN);
 
         // schedule a new scan every minute
-        scanJob = scheduler.scheduleWithFixedDelay(() -> {
-            startScan();
-        }, 0, 1, TimeUnit.MINUTES);
+        scanJob = scheduler.scheduleWithFixedDelay(this::startScan, 0, 1, TimeUnit.MINUTES);
+    }
+
+    protected void obtainIdentityAndPreSharedKey() {
+        TradfriGatewayConfig configuration = getConfigAs(TradfriGatewayConfig.class);
+
+        String identity = UUID.randomUUID().toString().replace("-", "");
+        String preSharedKey = null;
+
+        CoapResponse gatewayResponse;
+        String authUrl = null;
+        String responseText = null;
+        try {
+            DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder(new InetSocketAddress(0));
+            builder.setPskStore(new StaticPskStore("Client_identity", configuration.code.getBytes()));
+
+            DTLSConnector dtlsConnector = new DTLSConnector(builder.build());
+            CoapEndpoint authEndpoint = new CoapEndpoint(dtlsConnector, NetworkConfig.getStandard());
+            authUrl = "coaps://" + configuration.host + ":" + configuration.port + "/15011/9063";
+
+            CoapClient deviceClient = new CoapClient(new URI(authUrl));
+            deviceClient.setEndpoint(authEndpoint);
+
+            JsonObject json = new JsonObject();
+            json.addProperty(CLIENT_IDENTITY_PROPOSED, identity);
+
+            gatewayResponse = deviceClient.post(json.toString(), 0);
+
+            authEndpoint.destroy();
+            deviceClient.shutdown();
+
+            if (gatewayResponse.isSuccess()) {
+                responseText = gatewayResponse.getResponseText();
+                json = new JsonParser().parse(responseText).getAsJsonObject();
+                preSharedKey = json.get(NEW_PSK_BY_GW).getAsString();
+
+                if (isNullOrEmpty(preSharedKey)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Pre-shared key was not obtain successfully");
+                } else {
+                    logger.info("Recieved pre-shared key for gateway '{}'", configuration.host);
+                    logger.debug("Using identity '{}' with pre-shared key '{}'. Code can be discarded now", identity, preSharedKey);
+
+                    Configuration editedConfig = editConfiguration();
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_CODE, null);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_IDENTITY, identity);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_PRE_SHARED_KEY, preSharedKey);
+                    updateConfiguration(editedConfig);
+
+                    setupCoapClient();
+                }
+            } else {
+                logger.warn("Failed obtaining pre-shared key for identity '{}' (response code '{}', response text '{}')",
```

Comment:
You still have a WARNING here, why? This is not a bug situation, if we get here, is it?

Reply:
I think it is - because the response of the gateway was not a success - so authentication info can't be obtained and pairing can't continue.

(This may happen if the gateway declines the identity for whatever reason)

Reply:
E.g. because the passed security code is invalid...? Or what else happens in such a case?

Reply:
Not only the code is invalid, it can also be that the provided identity is already registered on the gateway or even I hit a situation where the identity was too long for the gateway (that's why I now do ```.replace("-", "");``` when creating it).

So - in such case - if the user can do anything depends on the particular situation. 
If the identity is already registered - re-initializing would work.
If the code is wrong - re-initializing would work after it's been updated
If the identity is too long - re-initializing would also work (because the new one would most probably be shorter)

Reply:
And you get exactly the same error code back in all those situations?

Reply:
Well - might be not the same (that's why the code is also put in the thing status) - but the fix usually is the same